from django_swagger_utils.drf_server.exceptions import BadRequest
from django_swagger_utils.drf_server.utils.decorator.interface_decorator \
    import validate_decorator

from ib_users.constants.otp_verification_constants import VerificationType
from ib_users.constants.res_status_constants import ResStatusConstants
from ib_users.interactors.DTOs.common_dtos import PhoneNumberDTO
from ib_users.interactors.exceptions.user_credentials_exceptions import \
    AccountWithPhoneNumberDoesntExistException
from ib_users.interactors.user_credentials.send_otp_phone_number_interactor \
    import \
    SendOTPPhoneNumberInteractor
from ib_users.interactors.verify_otp_through import VerifyThroughOTPInteractor
from ib_users.storages import UserAccountsSQLStorage
from ib_users.third_party.otp_service_impl import OTPServiceImpl
from ib_users.validators.base_validator import CustomException
from ib_users.validators.phone_number_validator_impl import \
    PhoneNumberValidatorImpl
from .validator_class import ValidatorClass


@validate_decorator(validator_class=ValidatorClass)
def api_wrapper(*args, **kwargs):
    request_data = kwargs['request_data']
    phone_number = request_data['phone_number']
    country_code = request_data['country_code']

    phone_number_dto = PhoneNumberDTO(
        phone_number=phone_number,
        country_code=country_code)

    storage = UserAccountsSQLStorage()
    verify_otp_interactor = VerifyThroughOTPInteractor(
        VerificationType.RESET_PASSWORD_WITH_OTP)
    interactor = SendOTPPhoneNumberInteractor(
        storage=storage, otp_service=OTPServiceImpl(verify_otp_interactor),
        phone_numbers_validator=PhoneNumberValidatorImpl())

    try:
        interactor.send_otp_to_phone_number(phone_number_dto=phone_number_dto)
    except CustomException as exception:
        raise BadRequest(message=exception.error_message,
                         res_status=exception.error_type)
    except AccountWithPhoneNumberDoesntExistException as exception:
        raise BadRequest(message=exception.message,
                         res_status=ResStatusConstants.ACCOUNT_WITH_THIS_PHONE_NUMBER_DOESNT_EXIST)